### PR TITLE
Fix data loss on secondary devices during device_put of sharded arrays

### PR DESCRIPTION
**Pull request by andersendsa**
closes: #36308

### CODE REVIEW

**Comment by andersendsa** [comment]
Hi @gnecula, @jakevdp  please let me know if there are any changes required or if this is good to merge

**Comment by Jimbob496** [comment]
Hi @andersendsa, thanks for working on this fix. I’ve tested your latest commit ```0b94dc80e``` by building JAX and the CUDA plugins from source but observe the same behaviour as in #36308
```
Expected 16.0 | observed 16.0
Expected 32.0 | observed 0.0
```
I include my installation below in case I made a mistake:

### Environment Setup
```
python -m venv jax_env
source jax_env/bin/activate
pip install -U pip
pip install setuptools wheel build numpy
```
### Building JAX
```
git clone https://github.com/andersendsa/jax.git
cd jax
git checkout 0b94dc8
nohup python build/build.py build \
    --wheels=jaxlib,jax-cuda-plugin,jax-cuda-pjrt \
    --cuda_version=13.0.0 \
    --cuda_major_version=13 \
    --cuda_compute_capabilities=sm_80 \
    --bazel_startup_options="--output_user_root=/tmp/bazel-cache-$USER" \
    --bazel_options="--local_ram_resources=HOST_RAM*.5" \
    --bazel_options="--local_cpu_resources=HOST_CPUS-2" \
    --bazel_options="--jobs=4" > jax_build.log 2>&1 &
pip install dist/*.whl
pip install -e .
pip install nvidia-cudnn-cu12
```
### Verification
```
python -c "import jax, jaxlib; print('JAX:', jax.__version__); print('JAXLIB:', jaxlib.__version__)"
```
```
JAX: 0.10.0.dev20260406+0b94dc80e
JAXLIB: 0.10.0.dev0+selfbuilt
```

**Comment by andersendsa** [comment]
> Hi @andersendsa, thanks for working on this fix. I’ve tested your latest commit `0b94dc80e` by building JAX and the CUDA plugins from source but observe the same behaviour as in #36308
> 
> ```
> Expected 16.0 | observed 16.0
> Expected 32.0 | observed 0.0
> ```
> 
> I include my installation below in case I made a mistake:
> 
> ### Environment Setup
> ```
> python -m venv jax_env
> source jax_env/bin/activate
> pip install -U pip
> pip install setuptools wheel build numpy
> ```
> 
> ### Building JAX
> ```
> git clone https://github.com/andersendsa/jax.git
> cd jax
> git checkout 0b94dc8
> nohup python build/build.py build \
>     --wheels=jaxlib,jax-cuda-plugin,jax-cuda-pjrt \
>     --cuda_version=13.0.0 \
>     --cuda_major_version=13 \
>     --cuda_compute_capabilities=sm_80 \
>     --bazel_startup_options="--output_user_root=/tmp/bazel-cache-$USER" \
>     --bazel_options="--local_ram_resources=HOST_RAM*.5" \
>     --bazel_options="--local_cpu_resources=HOST_CPUS-2" \
>     --bazel_options="--jobs=4" > jax_build.log 2>&1 &
> pip install dist/*.whl
> pip install -e .
> pip install nvidia-cudnn-cu12
> ```
> 
> ### Verification
> ```
> python -c "import jax, jaxlib; print('JAX:', jax.__version__); print('JAXLIB:', jaxlib.__version__)"
> ```
> 
> ```
> JAX: 0.10.0.dev20260406+0b94dc80e
> JAXLIB: 0.10.0.dev0+selfbuilt
> ```

hi it was a suggestion which i had taken by the gemini bot i hadnt tested locally i had tested the previous commit [acef914](https://github.com/jax-ml/jax/pull/36524/commits/acef9146917bebcd573bc4617d77e02e7d2f9aed)

**Review by yashk2810 — CHANGES_REQUESTED** [review]
Can you please add a test?

**Comment by andersendsa** [comment]
Hi @yashk2810  i have done the changes pls tell me if there are any other changes required or if this pr is good to merge

**Comment by andersendsa** [comment]
Hi @mattjj  could you please let me know if there are any changes required or if this is good to merge
